src/metaopt/algo/base.py

```diff
@@ -0,0 +1,186 @@
+
+from metaopt.core.utils import Factory
+
+log = logging.getLogger(__name__)
+
+
+@six.add_metaclass(ABCMeta)
+class BaseAlgorithm(object):
+    """Base class describing what an algorithm can do.
+
+    Notes
+    -----
+    We are using the No Free Lunch theorem's [1]_[3]_ formulation of an
+    `BaseAlgorithm`.
+    We treat it as a part of a procedure which in each iteration suggests a
+    sample of the parameter space of the problem as a candidate solution and observes
+    the results of its evaluation.
+
+    **Developer Note**: Each algorithm's complete specification, i.e.
+    implementation of its methods and hyperparameters of its own, lies in a
+    concrete algorithm class. Decorator (TODO) `declare_param` is provided in
+    `metaopt.algo.base` which enables developer to declare hyperparameters
+    of an algorithm implementation or wrapper.
+
+    References
+    ----------
+    .. [1] D. H. Wolpert and W. G. Macready, “No Free Lunch Theorems for Optimization,”
+       IEEE Transactions on Evolutionary Computation, vol. 1, no. 1, pp. 67–82, Apr. 1997.
+    .. [2] W. G. Macready and D. H. Wolpert, “What Makes An Optimization Problem Hard?,”
+       Complexity, vol. 1, no. 5, pp. 40–46, 1996.
+    .. [3] D. H. Wolpert and W. G. Macready, “No Free Lunch Theorems for Search,”
+       Technical Report SFI-TR-95-02-010, Santa Fe Institute, 1995.
+
+    """
+
+    def __init__(self, space, **hypers):
+        """Declare problem's parameter space and set up algo's hyperparameters.
+
+        Parameters
+        ----------
+        space : `metaopt.algo.space.Space`
+           Definition of a problem's parameter space.
+        hypers : dict
+           Tunable elements of a particular algorithm, a dictionary from
+           hyperparameter names to values.
+
+        """
+        log.debug("Creating Algorithm object of %s type with parameters:\n%s",
+                  type(self).__name__, hypers)
+        self._space = space
+        # Instantiate tunable parameters of an algorithm
+        for varname, hyper in six.iteritems(hypers):
+            # Check if tunable element is another algorithm
+            if isinstance(hyper, dict) and len(hyper) == 1:
+                subalgo_type = list(hyper)[0]
+                subalgo_hypers = hyper[subalgo_type]
+                if isinstance(subalgo_hypers, dict):
+                    hyper = OptimizationAlgorithm(subalgo_type,
+                                                  space, **subalgo_hypers)
+
+            setattr(self, varname, hyper)
+
+    @abstractmethod
+    def suggest(self, num=1):
+        """Suggest a `num`ber of new sets of parameters.
+
+        :param num: how many sets to be suggested.
+
+        .. note:: New parameters must be compliant with the problem's domain
+           `metaopt.algo.space.Space`.
+        """
+        pass
+
+    @abstractmethod
+    def observe(self, points, results):
+        """Observe evaluation `results` corresponding to list of `points` in
```

Comment:
@tsirif Do you think observe and judge could be merged? Otherwise I think they should be next to each other with a clear distinction between them in the doc. Beside the fact that one is on the results and the other on intermediate results, what is the difference?

Reply:
Answer in #44